### ISSUE TITLE
Global stat?

### ISSUE THREAD

**Issue by fredvs** [wp-xyz/corona#37]
What a useful project!
Surely the most clear and easy to use that I know.

Could it be possible to have also a global result?
I mean for all the planet and by continent?

Thanks.

Fre;D

**Comment by wp-xyz** [comment]
A global result will soon be available in the work-in-progress version in branch "mapping" since it loads the entire JHU data set in contrast to the current master version.

JHU does not provide a list with the assigment of countries to continents, and I am too lazy to create one on my own. So sorry about that...

**Comment by fredvs** [comment]
Hello Wp.

> A global result will soon be available in the work-in-progress version in branch "mapping" since it loads the entire JHU data set in contrast to the current master version

Ha, ok, I will follow it. 

> JHU does not provide a list with the assigment of countries to continents, and I am too lazy to create one on my own. So sorry about that...

No problem (but if you want I can take a look for it, just say me what you need).

**Comment by fredvs** [comment]
Here list of countries by continent in csv form:
[country-and-continent-codes-list-csv_csv.csv.zip](https://github.com/wp-xyz/corona/files/6038041/country-and-continent-codes-list-csv_csv.csv.zip)

**Comment by wp-xyz** [comment]
Thanks.

**Comment by wp-xyz** [comment]
The current version in the mapping branch now contains a top-level "world" node plus "continent" nodes, the countries have been assigned to the corresponding continent nodes according to your list. Thanks again.

**Comment by fredvs** [comment]
Congrats!

I will deeply look at it tonight.

Write you later.

Fre;D

**Comment by fredvs** [comment]
About the world-map and selection of countries with mouse.

What widget are you using for this?

**Comment by wp-xyz** [comment]
In TAChart there is a new polygon series which I copied and renamed into the corona sources so that the project can be compiled also with the release version. These polygons represent the features in the map (countries, states etc). Polygons can also be disconnected or can contain holes - I wrote a wiki about it recently. 

All the mouse handling is built into TAChart, there's not much to be done (except for tweaking the polygon series such that clicks inside the polygons are accepted, not only clicks on the polygon vertices.

The data for the polygons are provided by the TcGeoMap class which creates a reader for kml files. The structure is such that in principle also other geo formats (shp, geojson) can be integrated, but I did not go this far.

Maybe I'll add also maps for the continents. Depends whether I find decent files. The map files usually have a too fine resolution. For simplification I use the tool at https://mapshaper.org - this drops too-close polygon points; unlike other tools this one is clever enough to avoid generating gaps and overlaps between the countries. The re-saved, simplified maps are loaded into the application QGIS which I use only to write a kml file. The kml file sometimes must be edited manually, e.g. when the country names are not the same as in the JHU files. I also added a "GeoID" field which is unique to each country/state etc and is attached also to the polygon series. This way the series can be identifed by the GeoMap class.

**Comment by fredvs** [comment]
Wow.

Huh, are you the creator/contributor of TAChart?

(I did not yet deeply test corona app, I will do it tonight).

**Comment by fredvs** [comment]
Hello.

I try to compile corona-mapping with Linux 64 bit and Lazarus 2.0.10 but get this error (see picture)
![error-corona](https://user-images.githubusercontent.com/3421249/110852570-d2f51a00-82b2-11eb-82dd-f04decaff719.png)

**Comment by wp-xyz** [comment]
Oh yes. I usually work with trunk and only occasionally compile with other versions. I did not notice that TTreeView has many useful changes in trunk now. I added a class helper for TTreeNode to fix the issue for 2.0.10.

Yes, I am the current maintainer of TAChart.

**Comment by fredvs** [comment]
> I added a class helper for TTreeNode to fix the issue for 2.0.10.

Yep, compiles + runs perfectly now.
Super wow, impressive and very clean.

Huh, did you see the difference between continents ( max y of rich vs others)?
But this is a other discussion.

> Yes, I am the current maintainer of TAChart.

Congrats, excellent widget.

Fre;D

**Comment by fredvs** [comment]
> Huh, did you see the difference between continents (max y of rich vs others)?

Not really rich vs others, but Africa, Asia and Oceania does much  better than the others.
But, re-yes, it is a other discussion.

**Comment by fredvs** [comment]
Hello.

Perfect, deep test done, the earth-map works like charm, fluid and without problems.
There are very few little details here on Linux that dont look ok, for example the legends that are too long and dont fit on the screen, also for some check-boxes.

![Capture d’écran_2021-03-12_01-36-58](https://user-images.githubusercontent.com/3421249/110874789-a9e58100-82d4-11eb-9321-11be448a3036.png)

About the global stat for the all the world, is it a other ratio than for the continents because it dont seems to be the sum (or I miss something) ?

Anyway, superb work.

Fre;D
Also for some

**Comment by fredvs** [comment]
Last thing.

I see that Australia, China and some African countries are in white and the legend says _white = (no data)_
Is it true that for all that countries there are no data (or not reliable)?

**Comment by wp-xyz** [comment]
> There are very few little details here on Linux that dont look ok, for example the legends that are too long and dont fit on the screen, also for some check-boxes.

There should be splitters near such space-critical things so that you can adjust the size of the corresponding control as needed. Splitter positions are saved in the ini file so that you will be annoyed only once.

BTW it's painfully hard to write a self-layouting program which is correct for every theme and every widgetset even when anchoring and autosizing are active...

There is an issue with the y axis title of the time-series chart which is too long when also the map is displayed. I am planning to use a word-wrap function to split the title into several lines, depending on the space available

> Also the hint in the bottom chart is always visible (the timer dont hide it).

I am not sure about the chart hints, I think they are created by the application but by the chart and do not follow the conventions all the time. I added the hints to give the user an idea how to use the hidden parts of the program which do not have a place in the menu or toolbar. If you find the hints annoying you can always turn them off in the "Configuration" menu (not working at the moment for the map hints, but this will follow - or I'll find another place for the country info hints)

>  see that Australia, China and some African countries are in white and the legend says white = (no data)
> Is it true that for all that countries there are no data (or not reliable)?

I have a problem when the JHU files report "0" cases: does it mean "no data" or "zero value"? I decided to name it "no data" and have in mind that "no" could also mean "zero". 

But anyway, Australia and China did have non-zero values in the past. Did you notice the scrollbar underneath the map? It allows you to scroll back into the corona history. When you scroll back China about one year you'll see province Hubei getting a green color - which is a ridiculously low incidence compared to our current values. Do you remember the dramatic reports in these old days? When you select China/Hubei in the tree for the time-series chart and activate the "Date indicator" you see a purple line for the date displayed in the map.

------------

There was another question by you about the world data not being the sum of the continent values, but I can't find it any more (I hate the confusing notification system of github...). Well, the "new cases" and "cumulative cases" in the tree nodes for which JHU does not provide values were calculated exactly as the sum of the child node values, so this should be correct. But there are also the "normalized new cases" and these, in fact, do not add up. Here the new cases are are divided by the population of that region and multiplied by 100,000 -- i.e. this is the number of new cases normalized to a population of 100,000. And these numbers are not additive: a/b + c/d is not the same as (a+c)/(b+d). Suppose a region A having 100,000 inhabitants and 1 new case - the incidence (normalized new cases) is 1 here. In another region, B, there is also 1 new case, but there are only 1000 inhabitants -- the incidence is 100 (1/1000*100000). The naive sum for the combined region, A + B, is 1+100 but the total incidence of both regions would be 2 new cases for 100,000 + 1,000 = 101,000 inhabitants which corresponds to an incidence value slightly below 2.

**Comment by fredvs** [comment]
> BTW it's painfully hard to write a self-layouting program which is correct for every theme and every widgetset even when anchoring and autosizing are active...

Yes I know, same for me with MSEgui/fpGUI (and in my last projects I dont use anchor anymore for some widgets but code adapted for each OS).

> If you find the hints annoying you can always turn them off in the "Configuration" menu

Ooops, indeed, I did not see it, so ok, perfect.

> There was another question by you about the world data not being the sum of the continent values, but I can't find it any more

Huh, yes, I removed it after.  If you take a look at my previous image, Time series data is set to "Normalized cumulative" "Deaths" so I was thinking that the global stat was for this too.
But after I realized that it is the ratio deaths/100k inhabitants, so all is ok.

Fre;D

**Comment by fredvs** [comment]
Your new geotab is very impressive.
And it ask lot of things too.

For example, all that covid-deaths, what represent it as percentage of total deaths (all causes) in same period?

But, yes, I agree, there are lot of things to compare and a never-ended "I want more".

Anyway, nice project.

**Comment by wp-xyz** [comment]
>  If you take a look at my previous image, Time series data is set to "Normalized cumulative" "Deaths" so I was thinking that the global stat was for this too.

This is a problem with the current gui. I think it is not very intuitive at first glance that the upper groupbox refers to the map and the lower groupbox refers to the time series. I did try to move them into the chart panels, but this looks unbalanced. And there's a problem also with the treeview since it primarily determines which time series is shown, but it has also an effect on which map is seen.

**Comment by wp-xyz** [comment]
> For example, all that covid-deaths, what represent it as percentage of total deaths (all causes) in same period?

I think it would also be interesting to plot hospitality rate, count of intensive station beds available, and so on... Unfortunately the Johns Hopkins University files do not contain this information.

**Comment by fredvs** [comment]
> > If you take a look at my previous image, Time series data is set to "Normalized cumulative" "Deaths" so I was thinking that the global stat was for this too.
> 
> This is a problem with the current gui. I think it is not very intuitive at first glance that the upper groupbox refers to the map and the lower groupbox refers to the time series. I did try to move them into the chart panels, but this looks unbalanced. And there's a problem also with the treeview since it primarily determines which time series is shown, but it has also an effect on which map is seen.

Imho, the "Map Data" combo-box could be inside the map tab.
And maybe no more left panel but only the tabwidget in all the page and all what was in the left panel integrated into the time-serie panel.

Also, to reduce the length of legend, maybe use 10k instead of 10,000 (but this is detail).

> I think it would also be interesting to plot hospitality rate, count of intensive station beds available, and so on... 

Yes and maybe also to plot the new births in same period, this to relativize things.

**Comment by wp-xyz** [comment]
> Imho, the "Map Data" combo-box could be inside the map tab.
> And maybe no more left panel but only the tabwidget in all the page and all what was in the left panel integrated into the time-serie panel.

Normally this would be right, but the setup information for the map chart and the time series chart require a much different amount of space. When I do as you propose the treeview would not longer fit on the form when both charts are visible and the form height is 600 (that's my limit - I have VMs in which I cannot control the main window height, and then large forms are very difficult to handle)

An alternative would be to separate map and time series charts on different tabs of the pagecontrol. But I think it is very nice to have both of them on the same screen in particular when scrolling through the time history.

**Comment by fredvs** [comment]
> > Imho, the "Map Data" combo-box could be inside the map tab.
> > And maybe no more left panel but only the tabwidget in all the page and all what was in the left panel integrated into the time-serie panel.
> 
> Normally this would be right, but the setup information for the map chart and the time series chart require a much different amount of space. When I do as you propose the treeview would not longer fit on the form when both charts are visible and the form height is 600 (that's my limit - I have VMs in which I cannot control the main window height, and then large forms are very difficult to handle)
> 
> An alternative would be to separate map and time series charts on different tabs of the pagecontrol. But I think it is very nice to have both of them on the same screen in particular when scrolling through the time history.

Yes you are right, now after playing a few with it, it is ok (apart maybe the length of legends and checklistbox)
For the width-size of checklistbox, I know it is difficult to a adapt to the length of the biggest item.
Maybe there is a auto-size in LCL checklistbox?

**Comment by wp-xyz** [comment]
> apart maybe the length of [...] checklistbox

As I said there is a splitter at the left of the checklistbox and another splitter between the left parameter panel and the page control.

**Comment by fredvs** [comment]
> > apart maybe the length of [...] checklistbox
> 
> As I said there is a splitter at the left of the checklistbox and another splitter between the left parameter panel and the page control.

Ha, yes, ok, there is a  hidden spliter, I see it now and indeed the size can be done, perfect.
Imho, maybe the grip of the spliter should be a few more visible (lightly darker or with hatch).

**Comment by fredvs** [comment]
> Imho, maybe the grip of the spliter should be a few more visible (lightly darker or with hatch).

No, your hidden grip looks better sorry for the noise.

**Comment by fredvs** [comment]
> As I said there is a splitter at the left of the checklistbox and another splitter between the left parameter panel and the page control.

Ha, so much simpler, so for the first load of corona, just set the position of the listchekbox-spliter a few bigger for GTK2 vs win32.
This to have a nice first look.
Ok, I let you in peace and congrats for this impressive project (that gives lot of interrogations).
And yout TAMapChart is wonderful.

Fre;D
